### DISCUSSION TITLE
[AdonisJS 5] Error Access denied for user for driver mysq and mysq2

### DISCUSSION THREAD

**Discussion by andre-ab** [adonisjs/core#3539] [Help] [answered]
I have an error in the two MySQL drivers:
AdonisJs 5
Mysql 8
Debian 11

my settings:
**mysql**
![image](https://user-images.githubusercontent.com/10952239/152718356-b9fda756-3797-43f7-afb6-78e834f2a2f3.png)
Error:
![image](https://user-images.githubusercontent.com/10952239/152717691-45d80cf3-9bd6-40f3-8180-905931722ba6.png)

**mysql2**
![image](https://user-images.githubusercontent.com/10952239/152717520-c74c5024-a867-4921-9fe4-271231151941.png)
Error:
![image](https://user-images.githubusercontent.com/10952239/152717607-6fa4784c-d1d8-4177-8dd4-0ff5fb5f88e4.png)
Error:

I have access by express using knex in mysql and mysq2 drivers
my senttings express + knexjs
![image](https://user-images.githubusercontent.com/10952239/152717808-d3e67419-f063-4386-9c67-7380b87770fe.png)
Ok:
![image](https://user-images.githubusercontent.com/10952239/152718782-e0668a3a-9890-46b5-b44a-e2f5fb1976a3.png)

Mysq2
![image](https://user-images.githubusercontent.com/10952239/152718899-4f4f8298-6e01-4281-8299-342d32686dd1.png)
Ok
![image](https://user-images.githubusercontent.com/10952239/152718909-f1b96a80-2af4-4dd7-a194-925272cb9886.png)

I have access by Squirrel
![image](https://user-images.githubusercontent.com/10952239/152717993-e67aa49b-2b71-4726-a2ed-444e0c09b712.png)

**Comment by tomgobich** [comment, marked as answer]
`Env.get()` takes a lookup key as the first argument that it'll use to grab a value from the `.env` file. You can use the second argument for fallback values, though I'd recommend the below over that.

What you want to do is add the values you placed inside your MySQL config inside the .env file as the following:
```
DB_CONNECTION=mysql
MYSQL_HOST=localhost
MYSQL_PORT=3306
MYSQL_USER=aldisseia
MYSQL_PASSWORD=mid800
MYSQL_DB_NAME=mydb
```

The premise of environment variables is to keep things like this secret and outside of the code. The `.env` file shouldn't be committed to GitHub or any other repository. 

Once you've updated your `.env` file to the above, reset your MySQL config back to how it initially was:
```typescript
mysql: {
    client: 'mysql',
    connection: {
      host: Env.get('MYSQL_HOST'),
      port: Env.get('MYSQL_PORT'),
      user: Env.get('MYSQL_USER'),
      password: Env.get('MYSQL_PASSWORD', ''),
      database: Env.get('MYSQL_DB_NAME'),
    },
    migrations: {
      naturalSort: true,
    }, 
    healthCheck: false,
    debug: false,
  },
```
This will use the `MYSQL_HOST`, `MYSQL_PORT`, etc keys to look up the value within your `.env` and it'll use the matched value.

It's also recommended to add these to your `env.ts` file. You can think of this as a validation file for your environment variables (`.env`).

```typescript
DB_CONNECTION: Env.schema.string(),
MYSQL_HOST: Env.schema.string({ format: 'host' }),
MYSQL_PORT: Env.schema.number(),
MYSQL_USER: Env.schema.string(),
MYSQL_PASSWORD: Env.schema.string.optional(),
MYSQL_DB_NAME: Env.schema.string(),
```

You can read through the Environment Variables documentation for more information.
https://docs.adonisjs.com/guides/environment-variables

**Reply by andre-ab** [reply]
thanks for answering

**The error persists**
![image](https://user-images.githubusercontent.com/10952239/152798651-4c6fbbc2-1cb5-465b-ad45-4b4d9d655260.png)

**my routes.ts**
![image](https://user-images.githubusercontent.com/10952239/152798983-d85e9074-b146-43c2-b91b-ff1567f64cb2.png)

Error for mysql2:
![image](https://user-images.githubusercontent.com/10952239/152799090-282ec1a1-0379-4666-a8d7-11c3985a70e8.png)

Erro for mysql
![image](https://user-images.githubusercontent.com/10952239/152799836-1f4ddda9-c124-475d-a764-9483224a482e.png)

**Reply by tomgobich** [reply]
Please look at how your `Env.get` is being used compared to what I provided above. `Env.get('MYSQL_HOST')` maps to the MYSQL_HOST key inside the `.env` file. The way you have it you're trying to provide the value as the key, which won't work.

```typescript
mysql: {
    client: 'mysql',
    connection: {
      host: Env.get('MYSQL_HOST'), // <-
      port: Env.get('MYSQL_PORT'), // <- 
      user: Env.get('MYSQL_USER'), // <-
      password: Env.get('MYSQL_PASSWORD', ''), // <-
      database: Env.get('MYSQL_DB_NAME'), // <-
    },
    migrations: {
      naturalSort: true,
    }, 
    healthCheck: false,
    debug: false,
  },
```

**Reply by andre-ab** [reply]
**Thank you so much.**
is working now for mysql2, the error persists in mysql.

**Comment by galaczi** [comment]
You have to use the defined env variables with Env.get. Like:

host: Env.get('HOST')

**Reply by andre-ab** [reply]
None of them work
![image](https://user-images.githubusercontent.com/10952239/152803617-edae0a60-b79f-42a4-ae6a-f71d13ed5663.png)

**Reply by galaczi** [reply]
Check Tom's reply. You have to use only the name of the environment variable.

**Reply by andre-ab** [reply]
Thanks for answering.
it's working for mysql2.
error for mysq.
Thanks.

**Comment by andre-ab** [comment]
It's working. You can close as resolved.
"I couldn't find a way to close a thread as resolved."
Thank you all.

**Reply by WeyIin** [reply]
> It's working. You can close as resolved. "I couldn't find a way to close a thread as resolved." Thank you all.

You need to accept an answer which was helpful to be auto-closed.
